### ISSUE TITLE
[HOLD for payment 2023-04-27] [$1000] Edited autolink with several underscores contains incorrect target URL

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#16475]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Open [staging.new.expensify.com](http://staging.new.expensify.com/)
2. Open chat and send a message
3. Edit the message and replace it with a URL with several underscores (Example : https://www.facebook.com/hashtag/__main/?__eep__=6 )

## Expected Result:
Autolink URL from edited message should rendered as it is

## Actual Result:
One of the underscores in the hyperlink target replaced with em  and some parts of the hyperlink text is rendered as italic

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.2.88-1
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/43996225/227560332-ad23e94d-1b86-45a6-b738-97d17e625840.mp4

https://user-images.githubusercontent.com/43996225/227560396-6a187c83-3fff-4c97-afb9-5cb82803fff1.mp4

**Expensify/Expensify Issue URL:**
**Issue reported by:** @kerupuksambel
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1679633088263169

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0136dd0251d483d039</li>
        <li>Upwork Job ID: 1641797489981591552</li>
        <li>Last Price Increase: 2023-04-07</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by alex-mechler** [comment]
Yeah, this looks like a good ole regression, or just a spot that was somehow missed in that pr.

I don't think yours is a reproduction though @MitchExpensify, none of that text is italic.

**Comment by MitchExpensify** [comment]
Oh lol, yes you are right @alex-mechler, silly me! Would a next step be to reopen that issue seeing as the regression originated there, close this, and compensate @kerupuksambel for catching it again?

**Comment by alex-mechler** [comment]
I would leave this one open, since it has been several months this counts as a new / different regression

**Comment by MitchExpensify** [comment]
@kerupuksambel I'm unable o reproduce the italics, I wonder if its something to do with how you are copying and pasting the link, mind trying typing it out?

**Comment by kerupuksambel** [comment]
@MitchExpensify Sure thing! Here's my URL string I used : `https://www.facebook.com/hashtag/__main/?__eep__=6`. However, please make sure you send the message as an edit instead of a new message, so the steps should be : 

1. Send a sample message
2. Edit the message
3. Edit it to  `https://www.facebook.com/hashtag/__main/?__eep__=6`
4. Save it and examine the resulted (edited) string

That way I still able to reproduce the error.
![staging new expensify com_r_5122668032896954](https://user-images.githubusercontent.com/24503760/228404177-13d728ac-0c73-4cb1-af99-58d6bb67e5ac.png)

**Comment by MitchExpensify** [comment]
Thanks @kerupuksambel! Still unable to reproduce 

<img width="785" alt="image" src="https://user-images.githubusercontent.com/36425901/229135247-ae29e5bc-d1be-48da-b635-624cab8b2a5a.png">

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~0136dd0251d483d039

**Comment by MelvinBot** [comment]
Current assignee @MitchExpensify is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @fedirjh (`External`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @amyevans (`External`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by amyevans** [comment]
FYI I'll be OOO next week, feel free to reassign as needed to keep things moving @MitchExpensify

**Comment by eh2077** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Send a normal text message to chat like

```
example link
```

and then edit and replace the sent message with following link

```
https://www.faceless.com/hashtag/__main/?__eep__=6
```

Then some parts of the hyperlink text is rendered as italic.

### What is the root cause of that problem?

When saving link `https://www.faceless.com/hashtag/__main/?__eep__=6` to replace the sent message `example link`, the link is translated into [link markdown](https://github.com/Expensify/App/blob/265f8bfbf07c68e385c9578e77091873bcd8b80a/src/libs/actions/Report.js#L851)

```md
[https://www.faceless.com/hashtag/__main/?__eep__=6](https://www.faceless.com/hashtag/__main/?__eep__=6)
```

through [autolink-to-markdown](https://github.com/Expensify/App/blob/265f8bfbf07c68e385c9578e77091873bcd8b80a/src/libs/actions/Report.js#L830) and [anchor-html-to-markdown](https://github.com/Expensify/App/blob/265f8bfbf07c68e385c9578e77091873bcd8b80a/src/libs/actions/Report.js#L831) convertion in method [handleUserDeletedLinks](https://github.com/Expensify/App/blob/265f8bfbf07c68e385c9578e77091873bcd8b80a/src/libs/actions/Report.js#L825-L835). 

Then the [link markdown](https://github.com/Expensify/App/blob/265f8bfbf07c68e385c9578e77091873bcd8b80a/src/libs/actions/Report.js#L851) above is translated into following anchor html [here](https://github.com/Expensify/App/blob/265f8bfbf07c68e385c9578e77091873bcd8b80a/src/libs/actions/Report.js#L860) using [link-to-anchor-html regex rule](https://github.com/Expensify/expensify-common/blob/9cf047b9741d3c02820d515dccb9e0a45b6595f5/lib/ExpensiMark.js#L68-L92).

```html
<a href=\"https://www.faceless.com/hashtag/<em>_main/?__eep__=6\" target=\"_blank\" rel=\"noreferrer noopener\">https://www.faceless.com/hashtag/<em>_main/?__eep</em></em>=6</a>
```

The above anchor looks a little bit strange and we find that part of anchor tag text is translated into `em` tag by applying [italic-to-em-html regex rule](https://github.com/Expensify/expensify-common/blob/9cf047b9741d3c02820d515dccb9e0a45b6595f5/lib/ExpensiMark.js#L129-L142). This is because we support ['bold', 'strikethrough', 'italic'] [translations](https://github.com/Expensify/expensify-common/blob/9cf047b9741d3c02820d515dccb9e0a45b6595f5/lib/ExpensiMark.js#L403-L408) for text inside [] of link markdown text. That's why text part of anchor tag above includes `em` tag.

So the root cause is that we apply italic markdown translation for text part of link markdown. For example, we have link markdown

```md
[https://www.faceless.com/hashtag/__main/?__eep__=6](https://www.faceless.com/hashtag/__main/?__eep__=6)
```

We apply italic markdown translation for the text part(inside []), even if it is a url link.

### What changes do you think we should make in order to solve the problem?

To fix this issue, we should skip ['bold', 'strikethrough', 'italic'] [translations](https://github.com/Expensify/expensify-common/blob/9cf047b9741d3c02820d515dccb9e0a45b6595f5/lib/ExpensiMark.js#L403-L408) if the text part of link markdown is a url. 

We can use [URL_REGEX](https://github.com/Expensify/expensify-common/blob/9cf047b9741d3c02820d515dccb9e0a45b6595f5/lib/Url.js#L8) to test if the text part of link markdown is entirely a url. We can add a check [here](https://github.com/Expensify/expensify-common/blob/9cf047b9741d3c02820d515dccb9e0a45b6595f5/lib/ExpensiMark.js#L403-L408), like

```js
let linkText = match[1];
const urlRegex = new RegExp(`^${URL_REGEX}$`, 'gi');
if (!urlRegex.test(linkText)) {
    // `match[1]` contains the text inside []. Text is already escaped due to the rules
    // that precede the link rule (eg, codeFence, inlineCodeBlock, email).
    linkText = this.replace(match[1], {
        filterRules: ['bold', 'strikethrough', 'italic'],
        shouldEscapeText: false,
    });
}
```

`match[1]` is the text part of link markdown and is  the first capturing group of [regex](https://github.com/Expensify/expensify-common/blob/9cf047b9741d3c02820d515dccb9e0a45b6595f5/lib/ExpensiMark.js#L76-L79), see below picture

<img width="1154" alt="Screenshot 2023-04-01 at 10 19 43 AM" src="https://user-images.githubusercontent.com/117511920/229261416-a70eb10d-50d1-48f8-9df4-4583cf5f36c0.png">


### What alternative solutions did you explore? (Optional)

N/A

**Comment by tienifr** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
When editing the message and replace it with a URL with several underscores (Example : https://www.facebook.com/hashtag/__main/?__eep__=6 ), part of the link is rendered as italic

While if we send that link as a new message, the issue doesn't happen.

### What is the root cause of that problem?

There's inconsistency when we're parsing the edit message that contains auto link, compared to new message.

1. At the beginning when editing message, we have this text `https://www.facebook.com/hashtag/__main/?__eep__=6`
2. Then we apply the logic to remove the deleted links in [here](https://github.com/Expensify/App/blob/4ee21e95a38858df28e46caf54dc97e1bca9bc7a/src/libs/actions/Report.js#L851) which will convert the markdown to `[https://www.facebook.com/hashtag/__main/?__eep__=6](https://www.facebook.com/hashtag/__main/?__eep__=6)`
Basically we have converted all auto-link to regular-link markdown in this step, which is wrong.
3. Then we apply the other rules (except auto link) to the markdown in https://github.com/Expensify/App/blob/4ee21e95a38858df28e46caf54dc97e1bca9bc7a/src/libs/actions/Report.js#L860
The markdown is now parsed into HTML by the regular `link` rule, which will convert the `_` `_` markdown to italic in https://github.com/Expensify/expensify-common/blob/main/lib/ExpensiMark.js#L403-L408

Meanwhile when adding a new message, the `link` rule will never be applied because the auto link will be parsed by the `autolink` rule, not the regular `link` rule.

To sum up, we're incorrectly handling the parsing of the edited message (applying the `link` rule), causing the issue.

### What changes do you think we should make in order to solve the problem?

We need to make the parsing of the edit message and new message flows consistent.

When parsing to HTML, we should remove the deleted links only after applying all the rules in its normal order. (Rather than applying the `auto link` and handle edited links first, then apply the rest).

In https://github.com/Expensify/App/blob/main/src/libs/actions/Report.js, we should add a new function
```
const handleUserDeletedLinksInHtml = (newCommentText, html, originalHtml) => {
    const parser = new ExpensiMark();
    const markdownOriginalComment = parser.htmlToMarkdown(originalHtml).trim();
    const removedLinks = getRemovedMarkdownLinks(markdownOriginalComment, newCommentText);
    return removeLinksFromHtml(html, removedLinks);
};
```
(this is similar to the existing `handleUserDeletedLinks`, but for HTML)

Also the 
```
const removeLinksFromHtml = (html, links) => {
    let htmlCopy = html.slice();
    _.forEach(links, (link) => {
        const regex = new RegExp(`<(a)[^><]*href\\s*=\\s*(['"])${link}\\2(?:".*?"|'.*?'|[^'"><])*>([\\s\\S]*?)<\\/\\1>(?![^<]*(<\\/pre>|<\\/code>))`, 'gi');
        htmlCopy = htmlCopy.replace(regex, '$3');
    });
    return htmlCopy;
};
```
(this is similar to the existing `removeLinks`, but for HTML)

Then [this line](https://github.com/Expensify/App/blob/4ee21e95a38858df28e46caf54dc97e1bca9bc7a/src/libs/actions/Report.js#L860) needs to be changed to
```
        htmlForNewComment = parser.replace(textForNewComment);
        htmlForNewComment = handleUserDeletedLinksInHtml(textForNewComment, htmlForNewComment, originalCommentHTML);
 ```
 
 The above will make the rules apply order of `edit message` and `new message` consistent and will fix the issue.

### What alternative solutions did you explore? (Optional)
- We can remove some duplications by removing the logic to handle deleted links for markdown (since we already have ones for HTML), the markdown can be retrieved by converting back from our correct HTML instead.

### Result
Working well after the fix:
https://user-images.githubusercontent.com/113963320/229268119-b5be2553-3ae3-4c24-87ab-06d4080df573.mp4

**Comment by eh2077** [comment]
@fedirjh Thanks for reviewing proposals!

I just updated the RCA part of [my proposal](https://github.com/Expensify/App/issues/16475#issuecomment-1492798632).

**Comment by tienifr** [comment]
@fedirjh It's due to the `removeLinksFromHtml` mentioned in my implementation above has some issues, it's not escaping the `link` text passed into the Regex so if the `link` contains special character like `?`, it will be considered as a RegEx character rather than literal string character. So the `removeLinksFromHtml` will not find the link to delete.

To fix that we need to add a function to escape the link (see [here](https://stackoverflow.com/questions/3446170/escape-string-for-use-in-javascript-regex) for explanation on this function)
```
function escapeRegExp(string) {
    return string.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}
```

And use it like `escapeRegExp(link)` when passing the `link` in the RegEx in `removeLinksFromHtml`.

Working well now:

https://user-images.githubusercontent.com/113963320/229331432-6b476482-1f13-4f08-9fc8-d454c499dff6.mp4

**Comment by MitchExpensify** [comment]
> @MitchExpensify , @amyevans, While reviewing the proposals for the current issue, We noticed another related bug. The remove link feature not working properly with links that has special character. Should we address it in the current issue or a separate issue ?

Which is the "special character" being referenced here?

**Comment by MitchExpensify** [comment]
Got it! Thanks. I'll defer to @amyevans on whether that related issue can be fixed in this issue or if we should create a seperate one for it

**Comment by MelvinBot** [comment]
@amyevans @MitchExpensify @fedirjh this issue was created 2 weeks ago. Are we close to approving a proposal? If not, what's blocking us from getting this issue assigned? Don't hesitate to create a thread in #expensify-open-source to align faster in real time. Thanks!

**Comment by MelvinBot** [comment]
📣 It's been a week! Do we have any satisfactory proposals yet? Do we need to adjust the bounty for this issue? 💸

**Comment by MelvinBot** [comment]
@amyevans, @MitchExpensify, @fedirjh Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by MitchExpensify** [comment]
Hey @amyevans! Mind taking a peak at [this comment](https://github.com/Expensify/App/issues/16475#issuecomment-1493231732) to keep this trucking along? 🙇‍♂️

**Comment by tienifr** [comment]
@fedirjh Thanks for the review! I have 1 small concern regarding the suggested change in @eh2077 's [proposal](https://github.com/Expensify/App/issues/16475#issuecomment-1492798632).

I wonder in the use case where we want to highlight a specific part of the URL, for example like this [https://google.com/_path_](https://google.com/path), where the `path` is in italic, I added it in Github using `[https://google.com/_path_](https://google.com/path)`. If we added the suggested change then the formatting might not work.

**Comment by tienifr** [comment]
thanks everyone, @eh2077 I'll help with testing the PR once its up 🚀

**Comment by eh2077** [comment]
I drafted a PR [here](https://github.com/Expensify/App/pull/17451/files) FYI.

**Comment by MelvinBot** [comment]
@amyevans @MitchExpensify @fedirjh @eh2077 this issue is now 3 weeks old. There is one more week left before this issue <a href='https://stackoverflowteams.com/c/expensify/questions/15233/15234#15234'>breaks WAQ</a> and will need to go internal. What needs to happen to get a PR in review this week? Please create a thread in #expensify-open-source to discuss. Thanks!

**Comment by eh2077** [comment]
Hi @fedirjh , the PR for repo expensify-common is ready for review https://github.com/Expensify/expensify-common/pull/517, thank you 🙏

**Comment by MitchExpensify** [comment]
Invites sent on [the Upwork job](https://www.upwork.com/jobs/~0136dd0251d483d039) to @eh2077 and @fedirjh for eventual Upwork payment

**Comment by eh2077** [comment]
> Invites sent on [the Upwork job](https://www.upwork.com/jobs/~0136dd0251d483d039) to @eh2077 and @fedirjh for eventual Upwork payment

Hi @MitchExpensify , thanks for taking care of this while I should share the compensation with @tienifr as decided by C+ team, see @fedirjh 's comment https://github.com/Expensify/App/issues/16475#issuecomment-1506098018

**Comment by eh2077** [comment]
@fedirjh @amyevans The PR for App repo is ready for review https://github.com/Expensify/App/pull/17451. Please help to review it at your convenience, thanks!

**Comment by MelvinBot** [comment]
`Reviewing` label has been removed, please complete the "BugZero Checklist".

**Comment by MelvinBot** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.3.2-5 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/17451


If no regressions arise, payment will be issued on 2023-04-27. :confetti_ball:

After the hold period is over and BZ checklist items are completed, please complete any of the applicable payments for this issue, and check them off once done.
- [ ] External issue reporter
- [ ] Contributor that fixed the issue
- [ ] Contributor+ that helped on the issue and/or PR

As a reminder, here are the bonuses/penalties that should be applied for any External issue:
- Merged PR within 3 business days of assignment - 50% **bonus**
- Merged PR more than 9 business days after assignment - 50% **penalty**

**Comment by MelvinBot** [comment]
**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [ ] [@fedirjh] The PR that introduced the bug has been identified. Link to the PR:
- [ ] [@fedirjh] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
- [ ] [@fedirjh] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
- [ ] [@MitchExpensify] Determine if we should create a regression test for this bug.
- [ ] [@fedirjh] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.
- [ ] [@MitchExpensify] Link the GH issue for creating/updating the regression test once above steps have been agreed upon:

**Comment by MitchExpensify** [comment]
Made a calendar reminder to pay on the 27th 👍

**Comment by amyevans** [comment]
> A discussion in #expensify-bugs has been started : cc @amyevans Not sure this applies here.

Agreed, N/A.

The regression test proposal LGTM 👍

**Comment by MitchExpensify** [comment]
Everybody is paid out with bonus apart from @kerupuksambel (just waiting on you to accept the offer 😄)

**Comment by kerupuksambel** [comment]
@MitchExpensify thank you for the heads-up, I have accepted the offer

**Comment by MitchExpensify** [comment]
All done and new regression test requested
